Local time stepping (time steppers)

wthrowe opened the pull request:
This adds support to the AdamsBashforthN TimeStepper for local time stepping.  RungeKutta3 continues to only work with global time stepping, but implements the same interface.  This does not cover the driver and communication routines because we don't have the infrastructure for those in place yet.

wthrowe commented:
Known question: Where does LagrangePolynomial belong?

kidder commented on src/DataStructures/GeneralIndexIterator.hpp:
```diff
@@ -0,0 +1,67 @@
+/// Iterate over a generic "block" of coordinates.  The constructor
+/// takes a container of pairs defining the range (inclusive start,
+/// exclusive end) of each coordinate.  The utility function
+/// make_GeneralIndexIterator is provided for easier construction.
```
check doc (may need ())

kidder commented on src/DataStructures/GeneralIndexIterator.hpp:
```diff
@@ -0,0 +1,67 @@
+
+  /// Iterate over the coordinates of the current value.
+  //@{
+  iterator_type cbegin() const { return current_.cbegin(); }
```
noexcept everything that is noexcept

kidder commented on src/DataStructures/GeneralIndexIterator.hpp:
```diff
@@ -0,0 +1,67 @@
+
+/// Construct a GeneralIndexIterator
+template <typename Container>
+GeneralIndexIterator<std::decay_t<Container>> make_GeneralIndexIterator(
```
make_general_...

kidder commented on src/Numerical/Spectral/LagrangePolynomial.hpp:
```diff
@@ -0,0 +1,45 @@
+/// Evaluate the jth Lagrange interpolating polynomial with the given
+/// control points where j is the index of index_point.
+template <typename Iterator>
+double LagrangePolynomial(const Iterator& index_point,
```
lagrange_polynomial

kidder commented on src/Numerical/Spectral/LagrangePolynomial.hpp:
```diff
@@ -0,0 +1,45 @@
+/// Evaluate the jth (zero-indexed) Lagrange interpolating polynomial
+/// with the given control points.
+template <typename Iterator>
+double LagrangePolynomial(size_t j, double x,
```
snake_case

kidder commented on tests/Unit/Numerical/Spectral/Test_LagrangePolynomial.cpp:
```diff
@@ -0,0 +1,36 @@
+    for (size_t j = 0; j < control.size(); ++j) {
+      CHECK(LagrangePolynomial(j, control[i], control.begin(), control.end()) ==
+            approx(i == j ? 1. : 0.));
+      CHECK(LagrangePolynomial(std::next(control.begin(),
```
use + instead of next

kidder commented on src/Utilities/CachedFunction.hpp:
```diff
@@ -0,0 +1,59 @@
+template <typename Function, typename Map>
+class CachedFunction {
+ public:
+  using Input = typename Map::key_type;
```
input

kidder commented on src/Utilities/CachedFunction.hpp:
```diff
@@ -0,0 +1,59 @@
+class CachedFunction {
+ public:
+  using Input = typename Map::key_type;
+  using Output = typename Map::mapped_type;
```
output

kidder commented on src/Utilities/CachedFunction.hpp:
```diff
@@ -0,0 +1,59 @@
+/// \param map_args arguments to pass to the map constructor
+template <typename Input, template <typename...> class Map = std::unordered_map,
+          typename... MapArgs, typename Function, typename... PassedMapArgs>
+auto make_CachedFunction(Function function,
```
make_cached_function

kidder commented on src/Utilities/CachedFunction.hpp:
```diff
@@ -0,0 +1,59 @@
+          typename... MapArgs, typename Function, typename... PassedMapArgs>
+auto make_CachedFunction(Function function,
+                         PassedMapArgs... map_args) noexcept {
+  using Output = std::result_of_t<Function&(const Input&)>;
```
output

kidder commented on tests/Unit/Utilities/Test_CachedFunction.cpp:
```diff
@@ -0,0 +1,61 @@
+    return a < b;
+  };
+
+  // [make_CachedFunction_example]
```
check if need ///

kidder reviewed (changes requested):
for some of the commits

wthrowe committed:
Add structural comparison functor to Time

wthrowe committed:
Add GeneralIndexIterator

wthrowe committed:
Move history retirement logic to TimeSteppers

wthrowe committed:
Add AdamsBashforthN boundary coupling

wthrowe committed:
Add LagrangePolynomial

wthrowe committed:
Add TimeStepper boundary interface and RK3 impl

wthrowe committed:
Fix TimeStepper::update_u to match derived classes

wthrowe committed:
Add TimeStepper::needed_boundary_history

wthrowe committed:
Put time stepper test utilities in a namespace

wthrowe committed:
Add CachedFunction

wthrowe commented:
Addressed comments and clang-tidy, reset all the author dates so the commits show up in order in GitHub.  Changes to commits that have not been reviewed yet were immediately squashed.

One of the Travis builds had (and presumably will again have) a -Werror failure related to boost::optional that appears to just be the compiler's static analyzer not understanding the class.  I'm not sure of a way to fix that other than adding warning pragmas.

nilsdeppe commented on src/Time/Time.hpp:
```diff
@@ -57,6 +57,25 @@ class Time {
+  /// boundaries do not compare equal.  It is, however, much faster to
+  /// compute than the temporal ordering, so is useful when an
+  /// ordering is required, but the ordering does not have to be
+  /// physically meaningful.
```
"So this can be used as the key to a map"

nilsdeppe commented on src/Time/TimeSteppers/TimeStepper.hpp:
```diff
@@ -59,6 +62,35 @@ class TimeStepper : public Factory<TimeStepper> {
+  /// `std::vector<std::reference_wrapper<const FluxVars>>`.  These
+  /// values may be used to form a linear combination internally, so
+  /// `BoundaryVars` should have appropriate mathematical operators
+  /// defined to allow that.
```
`vector` -> neighbors, first value is self

`deque` -> time history

nilsdeppe commented on src/Time/TimeSteppers/RungeKutta3.hpp:
```diff
@@ -43,23 +48,64 @@ class RungeKutta3 : public TimeStepper::Inherit {
+    const std::vector<std::deque<std::tuple<Time, BoundaryVars, FluxVars>>>&
+        history,
+    const TimeDelta& time_step) const noexcept {
+  std::vector<std::reference_wrapper<const FluxVars>> coupling_args;;
```
remove extra `;`

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -98,6 +101,44 @@ class AdamsBashforthN : public TimeStepper::Inherit {
+    bool forward_in_time_;
+  };
+
+  /// Convert an iterator over history entries to one over their times.
```
one -> an iterator

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -98,6 +101,44 @@ class AdamsBashforthN : public TimeStepper::Inherit {
+    return boost::make_transform_iterator(it, get_time);
+  }
+
+  /// Find the step on a give side containing the given time.  Steps
```
give -> given

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+
+  const size_t num_sides = history.size();
+
+  // Start and end of the stepwe are trying to take
```
step we

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+        [target_order = target_order_s](const Iter& it) noexcept {
+          const auto it_end = std::next(it);
+          const auto it_start = std::prev(it, target_order - 1);
+          return TimeSteppers::AdamsBashforthN::get_coefficients(
```
reduce qualifiers?

fmahebert commented on src/Time/Time.hpp:
```diff
@@ -58,12 +58,13 @@ class Time {
-  /// ordering is required, but the ordering does not have to be
-  /// physically meaningful.
+  /// just looking at the class members.  This is only intended for
+  /// use as the comparitor in a map.  The returned ordering does not
```
comparator

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+    // current coupling evaluation.
+
+    // Latest of start of current step (otherwise computed in a
+    // preious call) and the times of the side data being coupled
```
previous call

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+        std::lower_bound(smallest_scale_times_begin, all_times.cend(),
+                         coupling_end_time, simulation_less);
+
+    // Step on each side containing the boundary step we are currently
```
"The step on each side"

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+    for (auto smallest_scale_time = smallest_scale_times_begin;
+         smallest_scale_time != smallest_scale_times_end;
+         ++smallest_scale_time) {
+      // Advance any side step where we've moved into the next one.
```
Advance the step of any side where we've moved into the next step

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+          if (ab_term == 0) { break; }
+
+          // Makes an iterator with a map to give time as a double.
+          const auto make_lagrange_iterator = [&](const HistoryIterator& it) {
```
only capture what's needed

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
+            boost::optional<BoundaryVars>(deriv_coef * coupling(args));
+      }
+    }
+  }
```
add namespace style comments at end of long for loops

nilsdeppe commented on tests/Unit/Time/TimeSteppers/Test_AdamsBashforthN.cpp:
```diff
@@ -74,3 +79,281 @@ SPECTRE_TEST_CASE("Unit.Time.TimeSteppers.AdamsBashforthN.Factory",
+constexpr double c21 = 0.805454101952822;
+constexpr double c22 = 0.825876851406978;
+
+double quartic_side1(double x) { return c10 + x * c11; }
```
how are these related to each other?

nilsdeppe commented on tests/Unit/Time/TimeSteppers/Test_AdamsBashforthN.cpp:
```diff
@@ -74,3 +79,281 @@ SPECTRE_TEST_CASE("Unit.Time.TimeSteppers.AdamsBashforthN.Factory",
+  double y = 0.;
+  while (t < slab.end()) {
+    for (size_t side = 0; side < 3; ++side) {
+    history[side].emplace_back(t, NCd(unused),
```
indent

nilsdeppe commented on tests/Unit/Time/TimeSteppers/TimeStepperTestUtils.hpp:
```diff
@@ -97,6 +95,8 @@ void integrate_test(const Stepper& stepper, const double integration_time,
     // This check needs a looser tolerance for lower-order time steppers.
     CHECK(y == approx(analytic(time.value())).epsilon(epsilon));
   }
+  // Make sure history is being cleaned up.
```
Mention that twenty is arbitrary, won't be exceeded in general cases

nilsdeppe commented on src/Time/Time.hpp:
```diff
@@ -58,12 +58,13 @@ class Time {
+  /// use as the comparitor in a map.  The returned ordering does not
+  /// match the time ordering and opposite sides of slab boundaries do
+  /// not compare equal.  It is, however, much faster to compute than
+  /// the temporal ordering, so is useful when an ordering is
```
so it's useful

wthrowe commented:
Trying to silence the warnings causing the Travis build failure with pragmas is somewhat problematic.  Some of them can be eliminated without trouble, but some of the warnings are nonlocal: they appear not in the time stepper but in the class used to represent the data.  For now that means some random function in the test, but this presumably means we will get warnings from DataVector or something when we try to use this in a simulation.

GCC bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635

kidder commented:
So, to be sure I understand, you are talking about the maybe-uninitialized warnings caused by the use of boost::optional.  Or are there other warnings?  Given the link you posted, I would suggest turning the maybe-uninitialized warnings off via compiler flag rather than pragmas if this is problematic

wthrowe commented:
That was what I was talking about, but I've realized I can avoid the optional by using make_arithmetic_value or the thing it's being replaced with, so silencing the warnings shouldn't be necessary.

wthrowe commented:
The fact that there are unsuppressable clang-tidy warnings that you can't fix is apparently intentional and not considered a bug: https://reviews.llvm.org/D26418#655058 (bullet 3).

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -354,75 +481,55 @@ BoundaryVars AdamsBashforthN::compute_boundary_delta(
-         smallest_scale_time != smallest_scale_times_end;
-         ++smallest_scale_time) {
+    // This loop computes the coefficient of the coupling evaluation
+    // as deriv_coef.
```
(i.e. the sum of the capital i s)

kidder commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -70,6 +70,98 @@ class AdamsBashforthN : public TimeStepper::Inherit {
       const TimeDelta& time_step) const noexcept;
 
+  /*!
+   * An explanation of the computation being performed by this
```
define k as the order of AB

wthrowe committed:
Silence clang-tidy warnings from ABNLTS test
